[gray-failure] Add simulation test that shows RemoteTLog issue

spraza opened the pull request:
In a scenario where a remote tLog has an outbound network issue, for example to a remote SS, then the gray failure algorithm currently does not detect and recover from this case. I am working on fixing this, but meanwhile sending the simulation test for review. In the simulation test, I have commented out the assert for now, and will enable it as part of the fix PR. 

# Code-Reviewer Section

The general pull request guidelines can be found [here](https://github.com/apple/foundationdb/wiki/FoundationDB-Commit-Process).

Please check each of the following things and check *all* boxes before accepting a PR.

- [ ] The PR has a description, explaining both the problem and the solution.
- [ ] The description mentions which forms of testing were done and the testing seems reasonable.
- [ ] Every function/class/actor that was touched is reasonably well documented.

## For Release-Branches

If this PR is made against a release-branch, please also check the following:

- [ ] This change/bugfix is a cherry-pick from the next younger branch (younger `release-branch` or `main` if this is the youngest branch)
- [ ] There is a good reason why this PR needs to go into a release branch and this reason is documented (either in the description above or in a linked GitHub issue)

spraza commented on tests/rare/ClogRemoteTlog.toml:
```diff
@@ -0,0 +1,32 @@
+processesPerMachine = 1
+machineCount = 20
+generateFearless = true
+minimumRegions = 2
```
FYI: Although `generateFearless` simulates HA mode with 4 DCs, it's possible all those DCs are in the same region (it happens in this PR with rocksdb storage engine). So setting `minimumRegions` to 2 explicitly since my test relies on 2 regions.

spraza commented on fdbserver/workloads/ClogRemoteTlog.actor.cpp:
```diff
@@ -0,0 +1,161 @@
+			}
+		}
+		TraceEvent("MaxSSDataLag").detail("SecondLag", maxSSLagSec).detail("SecondThreshold", self->lagThresholdSec);
+		// TODO (praza): Uncomment this assert when gray failure detection is improved to automatically fix this issue
+		// ASSERT(maxSSLagSec < self->lagThresholdSec);
```
I have confirmed that the test fails when the assert is uncommented.

jzhou77 commented on fdbserver/workloads/ClogRemoteTlog.actor.cpp:
```diff
@@ -0,0 +1,161 @@
+
+	ACTOR static Future<std::vector<IPAddress>> getRemoteSSIPs(Database db) {
+		state std::vector<IPAddress> ret;
+		Transaction tr(db);
+		tr.setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+		tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+		tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+		std::vector<std::pair<StorageServerInterface, ProcessClass>> results =
+		    wait(NativeAPI::getServerListAndProcessClasses(&tr));
+		for (auto& [ssi, p] : results) {
+			if (ssi.locality.dcId().present() && ssi.locality.dcId().get() == g_simulator->remoteDcId) {
+				ret.push_back(ssi.address().ip);
+			}
+		}
```
These txn read should be put into a `try...catch` block so that when encountering errors, the read can be retried.

jzhou77 commented on fdbserver/workloads/ClogRemoteTlog.actor.cpp:
```diff
@@ -0,0 +1,161 @@
+		TraceEvent("UnclogRemoteTLog").detail("SrcIP", remoteTLogIP).detail("DstIP", remoteSSIP);
+
+		// We only clog once, so this actor never finishes
+		wait(Never());
```
If this actor never finishes, then `workload()` can only exit because of a cancel due to timeout. Is this intentional?

spraza replied:
The cancel due to timeout (value is testDuration) is intentional. Overall intent is:

1. Run measurement and clog actors concurrently. 
2. If the measurement actor sees high SS lag, it asserts (will enable in feature PR). 
3. Clog actor's job is to just clog+unclog and then do nothing until end of test (based on timeout/testDuration value).

Any concerns with this approach?

Relatedly, I still want to ensure clog+unclog have already happened before end of test (based on timeout/testDuration). Theoretically, it looks like it could be possible that we are stuck somewhere and the test exits, but we haven't done clog+unclog. In this kind of situation, the test should fail. To fix this:

1. I can maintain explicit state that the clog+unclog happened, as well as we have enough measurement samples, and assert at test exit time that state value is as expected.
2. Try this approach: https://github.com/spraza/foundationdb/blob/392bad2bd31ba146e05641680929da2c10480bc5/fdbserver/workloads/ClogTlog.actor.cpp#L197-L201. But I am not sure how we pick `10` as in the workloadEnd calculation here: https://github.com/spraza/foundationdb/blob/392bad2bd31ba146e05641680929da2c10480bc5/fdbserver/workloads/ClogTlog.actor.cpp#L168. 

Any thoughts based on general practice of writing reliable simulation tests?

jzhou77 commented on fdbserver/workloads/ClogRemoteTlog.actor.cpp:
```diff
@@ -0,0 +1,161 @@
+		lagMeasurementFrequencySec = getOption(options, "lagMeasurementFrequencySec"_sr, 5);
+		clogInitDelaySec = getOption(options, "clogInitDelaySec"_sr, 5);
+		clogDurationSec = getOption(options, "clogDurationSec"_sr, 5);
+		lagThresholdSec = getOption(options, "lagThresholdSec"_sr, 5);
```
maybe try a larger value, say 60, to see if the commented out assertion still triggers?

spraza replied:
The current values in TOML file trigger the assert (if uncommented). I should update the default values here to reflect TOML file values, so that even if TOML file does not specify values, the defaults are good enough to reproduce the issue. Makes sense? Or did you have any other concern?

spraza commented:
Fyi: dropping this PR. The new test (with addressed feedback and other improvements) as well as the gray failure features are here: https://github.com/apple/foundationdb/pull/11717.
